src/accelerate/utils/other.py

```diff
@@ -75,6 +75,28 @@ def has_compiled_regions(module: torch.nn.Module) -> bool:
     return False
 
 
+def is_repeated_blocks(module: torch.nn.Module) -> bool:
+    """
+    Check whether the module is a repeated block, i.e. `torch.nn.ModuleList` of the same class. This is useful to
+    determine whether we should compile the module as a whole or keep looking for repeated blocks in its submodules.
+    """
+
+    return isinstance(module, torch.nn.ModuleList) and all(isinstance(m, module[0].__class__) for m in module)
+
+
+def has_repeated_blocks(module: torch.nn.Module) -> bool:
+    """
+    Check whether the module has repeated blocks, i.e. `torch.nn.ModuleList` of the same class. This is useful to
+    determine whether we should compile the module as a whole or keep looking for repeated blocks in its submodules.
+    """
+    if module._modules:
+        for submodule in module.modules():
+            if is_repeated_blocks(submodule):
+                return True
+
+    return False
```

Comment:
this helps reduce the number of regions by only compiling recursively when the module has repeated blocks, favoring compilation of bigger regions.